### PR TITLE
[#751] Add Import Static and Import Static Extension commands

### PR DESCRIPTION
**Pull request by Hanksha**
Add _Import Static_ and _Import Static Extension_ commands #751 

The commands are accessible from the `Refactoring > Import Static | Import Static Extension` menu of the Xtend editor or via shortcut `M1+SHIFT+M` and `M1+SHIFT+ALT+M`.

The  _Import Static_ command will add as static import the selected method (if possible) and refactoring all occurrences in the document so that it uses the static method call.

Example:

importstatic

See `org.eclipse.xtend.ide.tests.refactoring.importer.StaticMethodImporterTest` for the behaviors.

The  _Import Static Extension_ command will add as static extension import the selected method (if possible) and refactoring all occurrences in the document so that it uses the static extension method call. It means it will move the first argument and put it as target of the method call.

Example:

importstaticextension

See `org.eclipse.xtend.ide.tests.refactoring.importer.StaticExtensionMethodImporterTest` for the behaviors. It has two ignored tests that fail due to an issue with the `ImportOrganizer` which removes the import because a static explicit call doesn't mark it as used.

Signed-off-by: Vivien Jovet <vivien.jovet@gmail.com>

### CODE REVIEW

**Comment by Hanksha** [comment]
It seems I have some flaky tests, I tried to run the build locally via `mvn -f tycho-pom.xml clean install -Platest` but I'm getting this error `Unable to read repository at https://ci.eclipse.org/xtext/job/xtext-lib/job/master/lastStableBuild/artifact/build/p2-repository/content.xml. Received fatal alert: handshake_failure`

**Comment by cdietrich** [comment]
Which java version is used in maven ?

**Comment by Hanksha** [comment]
I'm using 3.6.0

**Comment by cdietrich** [comment]
Java version not maven version

**Comment by Hanksha** [comment]
Edited the comment, pressed enter my mistake.

**Comment by cdietrich** [comment]
That version mit be too old for modern tls connections

**Comment by cdietrich** [comment]
You should also try to run your test only by going into the test project and run

mvn clean install -Dtest=TestClassName

**Comment by Hanksha** [comment]
So I should try with a more recent java 8 version or java 9 ?

**Comment by cdietrich** [comment]
You should use the latest java 8. If you have concerns with Oracle Licensing you can Have Ooenjdk from the adoptopenjdk project

**Comment by Hanksha** [comment]
Using adoptopenjdk8 fixed it, thanks.

**Comment by cdietrich** [comment]
hmmm i did have a look at the code and have no clue why its failing. maybe we need to run this with some logging to see whats happening

**Comment by cdietrich** [comment]
one thing that looks wired to me is the file names

```
def void testSingleStaticMethod() {
        'test.Test.xtend'.
```

guess that should be `test/Test.xtend`
am not sure if this is related to the problem.

**Comment by cdietrich** [comment]
there are also some other invalid code places like

```
val i = 0
                    Other.doSo|mething(i)
                    Other.doSomething(i++)
```

**Comment by Hanksha** [comment]
You're right, I'll double check the tests.

**Comment by Hanksha** [comment]
I changed it to `test/Test.xtend` but a test that didn't fail before is now failing on jenkins (works locally).

**Comment by cdietrich** [comment]
I cloud not reproduce this locally so far. Will try on Monday with xvnc

**Comment by cdietrich** [comment]
the test / impl code does

``` val resource = document.priorityReadOnly[it]```
and then works with the resource. this looks bogus to me.
imho all the work on resource should happen within the "work"

**Comment by cdietrich** [comment]
@Hanksha i did some experiments and doing the 
`resourceCopier.loadIntoNewResourceSet(it)`
seems to help
https://github.com/eclipse/xtext-xtend/commit/2c127d714954bb93c39068205cd0645344e39a83
https://ci.eclipse.org/xtext/job/xtext-xtend/job/Hanksha-issue751/

**Comment by Hanksha** [comment]
@cdietrich Ok I'll try with that.

File: org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend [hunk 1]
```diff
@@ -527,6 +529,14 @@ class XtendUiModule extends AbstractXtendUiModule {
 	def Class<? extends SyntaxColoringPreferencePage> bindSyntaxColoringPreferencePage() {
 		return XtendSyntaxColoringPreferencePage
 	}
+	
+	def Class<? extends StaticMethodImporter> bindStaticMethodImporter() {
+	    return StaticMethodImporter
```
Comment (cdietrich) [inline org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: why is this needed?

Comment (Hanksha) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: Is it not needed to be able to inject it ? Or does it automatically consider it a singleton?

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: its not really needed but it helps. singletons require a `@Singleton` annotation

Comment (Hanksha) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: So do you prefer to remove it here or annotate `StaticMethodImporter` and `StaticExtensionMethodImporter` classes with `@Singleton`?

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: you did me got wrong. that you did is not a singleton binding.

Comment (Hanksha) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: I understood that, my question is, should I remove it from here then?

Comment (cdietrich) [reply org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/XtendUiModule.xtend]: no you can keep it

**Comment by Hanksha** [comment]
@cdietrich Looks like the tests are not flaky anymore.

**Comment by cdietrich** [comment]
yes the copy of the resources seems to have solved that issues. will try to find some time to go through the impl and tests later this week

**Comment by Hanksha** [comment]
Alright, thanks for the help.

**Comment by cdietrich** [comment]
Thx for the Contribution @Hanksha
